### PR TITLE
[temp] feat: add detail taxonomy page

### PR DESCRIPTION
**Pull request by rpenido**
## Supporting Information
- Closes https://github.com/openedx/modular-learning/issues/138
- Depends on:
  - https://github.com/openedx/frontend-app-course-authoring/pull/645
____
Private-ref: [FAL-3529](https://tasks.opencraft.com/browse/FAL-3529)

### CODE REVIEW

File: src/index.jsx
Comment (pomegranited) [inline src/index.jsx]: Is this necessary? I wasn't able to trigger it by visiting a non-existent page, so I think this part should be removed.
```suggestion
```

File: src/taxonomy/api/hooks/api.js
Comment (pomegranited) [inline src/taxonomy/api/hooks/api.js]: We shouldn't be hard-coding `page_size` into the API URL here; let the backend determine the default page size.

```suggestion
  `api/content_tagging/v1/taxonomies/${taxonomyId}/tags/`,
```

If we need the page_size to be configurable/modifiable by the user, then we can add it to Paragon's table component.

File: src/taxonomy/taxonomy-detail/TagListTable.jsx
Comment (pomegranited) [inline src/taxonomy/taxonomy-detail/TagListTable.jsx]: From https://github.com/openedx/modular-learning/issues/138

> Do not implement Keyword filtering for now - that will come in https://github.com/openedx/modular-learning/issues/128

So please remove the filter-related settings here? We'll add them when we're ready for them.

```suggestion
      isSortable
```

File: src/taxonomy/taxonomy-detail/TagListTable.jsx
Comment (pomegranited) [inline src/taxonomy/taxonomy-detail/TagListTable.jsx]: Missing a few pagination-related settings here (ref: [Paragon: DataTable](https://paragon-openedx.netlify.app/components/datatable/)):
```suggestion
      isPaginated
      manualPagination
      pageCount={tagList.numPages}
      fetchData={fetchData}
```

^ needs a `fetchData` method implemented to make the next/prev page buttons fetch the next page of data, see the example under [DataTable#paginated-selection](https://paragon-openedx.netlify.app/components/datatable/#backend-filtering-and-sorting).

File: src/taxonomy/taxonomy-detail/TaxonomyDetailSideCard.jsx
Comment (pomegranited) [inline src/taxonomy/taxonomy-detail/TaxonomyDetailSideCard.jsx]: Please omit this until we have copyright fields to show:
```suggestion
```

File: src/taxonomy/taxonomy-card/TaxonomyCard.jsx
Comment (pomegranited) [inline src/taxonomy/taxonomy-card/TaxonomyCard.jsx]: If you remove `isClickable` here, then the stretched-link be what gets followed when you click the card, or when you tab to the card and hit "enter".

File: src/taxonomy/taxonomy-card/TaxonomyCard.scss
Comment (pomegranited) [inline src/taxonomy/taxonomy-card/TaxonomyCard.scss]: This change looks like it belongs on https://github.com/openedx/frontend-app-course-authoring/pull/645, not here?

**Review by pomegranited — CHANGES_REQUESTED** [review]
I've run this on my devstack and it's looking great so far, @rpenido !

I've left a few suggestions, mainly around how the DataTable component is configured.

And as you've noted, there's some copy-paste code in here which should be factored out and moved up under `src/taxonomy` somewhere.

File: src/index.jsx
Comment (pomegranited) [inline src/index.jsx]: There's a routing bug that occurs if you follow these steps:

1. Visit http://localhost:2001/taxonomy-list/   <-- note trailing slash
2. Click on one of the taxonomies there, and note the URL looks like this: `http://localhost:2001/taxonomy-list/taxonomy-list/<pk>`

I'm not sure what the fix is, but I recommend looking at how the other nested routes are written and imitating that.

Comment (rpenido) [reply src/index.jsx]: Thank you for getting that!

The react-router was updated (https://github.com/openedx/frontend-app-course-authoring/pull/519) to V6. I will merge this update in the [FAL-3528](https://github.com/open-craft/frontend-app-course-authoring/tree/chris/FAL-3528-taxonomy-export-menu) branch and rebase here before fixing it!

Comment (rpenido) [reply src/index.jsx]: Fixed!

File: src/taxonomy/taxonomy-card/TaxonomyCard.jsx
Comment (rpenido) [reply src/taxonomy/taxonomy-card/TaxonomyCard.jsx]: The `isClickable` adds a little hover effect to the card.
I think the following (591604cfc692e0f20fa734be095f601589c3a2f6) is the right way to do it (without the need of the `.stretched-link` class).

File: src/taxonomy/taxonomy-card/TaxonomyCard.scss
Comment (rpenido) [reply src/taxonomy/taxonomy-card/TaxonomyCard.scss]: I needed this because the `stretched-link` was over the card menu and the tooltip from the system-defined badge. It was fixed using the clickable card (591604cfc692e0f20fa734be095f601589c3a2f6).

File: src/taxonomy/taxonomy-detail/TagListTable.jsx
Comment (rpenido) [reply src/taxonomy/taxonomy-detail/TagListTable.jsx]: Done 3d6b0dfe53b817861038e2758aa405cc9b182066

File: src/index.jsx
Comment (rpenido) [reply src/index.jsx]: Removed

File: src/taxonomy/api/hooks/api.js
Comment (rpenido) [reply src/taxonomy/api/hooks/api.js]: Added backend pagination: 3d6b0dfe53b817861038e2758aa405cc9b182066

File: src/taxonomy/taxonomy-detail/TagListTable.jsx
Comment (pomegranited) [inline src/taxonomy/taxonomy-detail/TagListTable.jsx]: Calling both `useTagListDataStatus` and `useTagListDataResponse` is unnecessary here -- if you watch the Network tab of your browser Inspector, you can see that we hit the taxonomy REST APIs twice every time we want to load a page of tags.

Both of those methods call the api.js's [`useTagListData`](https://github.com/open-craft/frontend-app-course-authoring/pull/5/files#diff-51e2c0d3363c83c5d167493ee852ee13f16564ad2c70189c74d66f8adaca8cd6R54) (and hit the REST API), they just return different slices of the response object, so it would be better to combine these into a single selector method.

File: src/taxonomy/taxonomy-detail/TaxonomyDetailPage.jsx
Comment (pomegranited) [inline src/taxonomy/taxonomy-detail/TaxonomyDetailPage.jsx:27]: Similar comment here as [with the tag list](https://github.com/open-craft/frontend-app-course-authoring/pull/5/files#diff-8e2e37b39ab5b552b68cc89e7dc30ec0789817c2f142fa3bd8d25fa61f8f7e33R17) -- can we avoid hitting the API twice here?

File: src/taxonomy/taxonomy-detail/TagListTable.jsx
Comment (pomegranited) [reply src/taxonomy/taxonomy-detail/TagListTable.jsx]: I haven't dug into figure out why, but when you hit the prev/next buttons on the table here, we hit both these REST APIs. But we should only be hitting the 2nd `/tags/` endpoint.

1. `/api/content_tagging/v1/taxonomies/<pk>/`
2. `/api/content_tagging/v1/taxonomies/<pk>/tags/`

File: src/index.jsx
Comment (pomegranited) [reply src/index.jsx]: Excellent, great fix @rpenido !

And I like how you pulled the common header/footer into a single `TaxonomyLayout` wrapper, that's perfect.

**Review by pomegranited — CHANGES_REQUESTED** [review]
These latest changes look great, @rpenido ! I found some issues with the REST API being hit more than we need, but your addition of `TaxonomyLayout`, and your taxonomy-detail components look great.

File: src/taxonomy/taxonomy-detail/TagListTable.jsx
Comment (rpenido) [reply src/taxonomy/taxonomy-detail/TagListTable.jsx]: Hi, @pomegranited!

This is my first time using `React-Query,` and I was also confused why multiple selectors are recommended by [our docs](https://openedx.atlassian.net/wiki/spaces/BPL/pages/3852632086/Typescript+and+React-Query+for+api+usage.#3%3A-Define-Selectors-for-the-API-hooks), afraid that it will lead to multiple API calls.

But the `React-Query` caches the API call result, so multiple calls are not supposed to happen, even using the current approach.

Could you check if the first call is a pre-flight one (method: `OPTION`)?
image
If it is, we are good, and everything works as expected!

File: src/taxonomy/taxonomy-detail/TaxonomyDetailPage.jsx
Comment (rpenido) [reply src/taxonomy/taxonomy-detail/TaxonomyDetailPage.jsx:27]: Ditto

File: src/taxonomy/taxonomy-detail/TagListTable.jsx
Comment (rpenido) [reply src/taxonomy/taxonomy-detail/TagListTable.jsx]: This is also related to `React-Query`. If we refocus the window or [other conditions](https://tanstack.com/query/v4/docs/react/guides/important-defaults), we get a fetch in the background. This is the default behavior. We could tune this application-wise (the same happens in the TaxonomyList) if we think this doesn't suit us.

I was not able to reproduce the issue here. Maybe a windows refocus (not prev/next) triggered the call?

This is from the react-query docs link above:
> If you see a refetch that you are not expecting, it is likely because you just focused the window and TanStack Query is doing a [refetchOnWindowFocus](https://tanstack.com/query/v4/docs/react/guides/window-focus-refetching). During development, this will probably be triggered more frequently, especially because focusing between the Browser DevTools and your app will also cause a fetch, so be aware of that.

Comment (bradenmacdonald) [reply src/taxonomy/taxonomy-detail/TagListTable.jsx]: Yeah, don't worry about extra fetch calls seen, it's just checking if the server data has changed, so it can automatically refresh the UI if needed. If the data hasn't changed, nothing will happen.

File: src/taxonomy/taxonomy-detail/TagListTable.jsx
Comment (bradenmacdonald) [reply src/taxonomy/taxonomy-detail/TagListTable.jsx]: Yep, react-query should always be caching the result of each API call, so you can call each endpoint as many times as you want, from different components at different levels, and use different parts of the result, but it should only result in a single API call*. You do have to make sure the URL is exactly the same though, and if there's anything else like user ID that affects the result it should be in the [query key](https://tanstack.com/query/v4/docs/react/guides/query-keys).

* But a single call can be two if you include the preflight, as @rpenido points out.

**Comment by bradenmacdonald** [comment]
Here is the PR that will merge soon and affect how taxonomy data is returned from the API - hopefully making it more flexible for you: https://github.com/openedx/edx-platform/pull/33553

File: src/taxonomy/TaxonomyLayout.jsx
Comment (pomegranited) [inline src/taxonomy/TaxonomyLayout.jsx]: nit: styles belong in `.scss` files?

File: src/taxonomy/taxonomy-card/TaxonomyCardMenu.jsx
Comment (pomegranited) [inline src/taxonomy/taxonomy-card/TaxonomyCardMenu.jsx:33]: I'm able to use the Export menu item -- clicking it takes me to the Taxonomy detail page.

So I think we need to run `e.preventDefault()` on the `MenuItem` onClick event below too to prevent the click event from bubbling up to the Card. And since there will be more menu items eventually, I think it's best to do it in the `onClickItem` method above, e.g

```javascript
const onClickItem = (e, menuName) => {
    e.preventDefault();
    setMenuIsOpen(false);
    onClickMenuItem(menuName);
  };
```

invoked below like this:

```javascript
<MenuItem className="taxonomy-menu-item" onClick={(e) => onClickItem(e, 'export')}>
```

**Review by pomegranited — APPROVED** [review]
Hi @rpenido there's a bug with the Export menu item, and I noted one optional nit.

But everything else is working great, so I'll approve pending that fix, and then it's ready for upstream review.

:+1:

- [x] I tested this on my devstack.
- [x] I read through the code
- [x] I checked for accessibility issues by using my keyboard to navigate the new and modified components.
- [x] <strike>Includes documentation</strike> N/A
- [x] Commit structure follows [OEP-0051](https://github.com/edx/open-edx-proposals/pull/189)

File: src/taxonomy/TaxonomyLayout.jsx
Comment (rpenido) [reply src/taxonomy/TaxonomyLayout.jsx]: Done: 6a46c29a869bf33eb748b7ef38ab9d0ae5201f38

File: src/taxonomy/TaxonomyLayout.scss
Comment (xitij2000) [inline src/taxonomy/TaxonomyLayout.scss]: This is equivalent to bg-light-400, so you can avoid the custom css.

File: src/taxonomy/TaxonomyLayout.test.jsx
Comment (xitij2000) [inline src/taxonomy/TaxonomyLayout.test.jsx:48]: I don't think this test is adding anything. There is no logic in this component so I think this file can be entirely skipped.

File: src/taxonomy/taxonomy-detail/TagListTable.jsx
Comment (xitij2000) [inline src/taxonomy/taxonomy-detail/TagListTable.jsx]: I think you could use the [Paragon Spinner instead](https://paragon-openedx.netlify.app/components/spinner/)

File: src/taxonomy/TaxonomyLayout.scss
Comment (rpenido) [reply src/taxonomy/TaxonomyLayout.scss]: Thank you! Fixed here: a12f2621fd4f9288fd0a610aac40143793aabcb2

File: src/taxonomy/TaxonomyLayout.test.jsx
Comment (rpenido) [reply src/taxonomy/TaxonomyLayout.test.jsx:48]: It only helps the code coverage. If this metric is not an issue here, let me know, and I remove it!

File: src/taxonomy/taxonomy-detail/TagListTable.jsx
Comment (rpenido) [reply src/taxonomy/taxonomy-detail/TagListTable.jsx]: This code is outdated. I'm using the Loading status from the DataTable

**Comment by bradenmacdonald** [comment]
@rpenido @xitij2000 What's the status of this PR? And how does it relate to https://github.com/openedx/frontend-app-course-authoring/pull/655 ?

**Comment by rpenido** [comment]
> @rpenido @xitij2000 What's the status of this PR? And how does it relate to [openedx#655](https://github.com/openedx/frontend-app-course-authoring/pull/655) ?

- The branch of this feature is based on https://github.com/openedx/frontend-app-course-authoring/pull/645 branch.
- The https://github.com/openedx/frontend-app-course-authoring/pull/655 is against `master` and is the official PR.

This is a temp PR against that branch to allow us to review the changes before https://github.com/openedx/frontend-app-course-authoring/pull/645 is merged.

I'm waiting for https://github.com/openedx/frontend-app-course-authoring/pull/645 approval to rebase this branch and bring the improvements from the review to here.

**Review by pomegranited — APPROVED** [review]
:+1: Thanks for that @rpenido ! This is good for upstream review now. CC @bradenmacdonald 

- [x] I tested this using the REST API on my devstack
- [x] I read through the code
- [x] <strike>I checked for accessibility issues</strike> N/A
- [x] Includes documentation
- [x] Commit structure follows [OEP-0051](https://github.com/edx/open-edx-proposals/pull/189)

**Comment by bradenmacdonald** [comment]
I guess this is blocked on https://github.com/openedx/frontend-app-course-authoring/pull/645 now ? Once that PR is merged, please open a new PR for this on the upstream repo.

**Comment by rpenido** [comment]
> I guess this is blocked on [openedx#645](https://github.com/openedx/frontend-app-course-authoring/pull/645) now ? Once that PR is merged, please open a new PR for this on the upstream repo.

Sure!
The PR on upstream is here: https://github.com/openedx/frontend-app-course-authoring/pull/655. I will rebase as soon as we get https://github.com/openedx/frontend-app-course-authoring/pull/645 merged.

File: src/taxonomy/data/types.mjs
Comment (pomegranited) [inline src/taxonomy/data/types.mjs]: These changes aren't needed here (artifact of rebase?)

File: src/taxonomy/tag-list/data/selectors.js
Comment (ChrisChV) [inline src/taxonomy/tag-list/data/selectors.js]: @rpenido From what I understand from [here](https://github.com/openedx/frontend-app-course-authoring/pull/645#discussion_r1377342836), these would not be considered selectors in the context of the other MFE modules. To avoid colliding with that context is why I have placed functions like this in `apiHooks.js`. But you can wait until https://github.com/openedx/frontend-app-course-authoring/pull/645 is approved to make those changes

File: src/taxonomy/taxonomy-detail/TaxonomyDetailMenu.jsx
Comment (ChrisChV) [inline src/taxonomy/taxonomy-detail/TaxonomyDetailMenu.jsx:1]: @rpenido In order to follow the [ADR](https://github.com/openedx/frontend-app-course-authoring/blob/master/docs/decisions/0002-feature-based-application-organization.rst#file-and-directory-naming), we need an `index.js` on this `taxonomy-detail` sub-module for the public interface

**Review by ChrisChV — APPROVED** [review]
@rpenido Looks good :+1: Check my comments, but I think you can wait until https://github.com/openedx/frontend-app-course-authoring/pull/645 is merged to do any change. 

- [X] I tested this: I follow the testing instructions on https://github.com/openedx/frontend-app-course-authoring/pull/655
- [X] I read through the code and considered the security, stability and performance implications of the changes.
- [X] I tested that the UI can be used with a keyboard only (tab order, keyboard controls).
- [X] Includes tests for bugfixes and/or features added.
- [X] Includes documentation

File: src/taxonomy/taxonomy-detail/TaxonomyDetailSideCard.jsx
Comment (pomegranited) [inline src/taxonomy/taxonomy-detail/TaxonomyDetailSideCard.jsx]: nit typo:
```suggestion
      <Card.Section title={intl.formatMessage(messages.taxonomyDetailsDescription)}>
```

**Comment by pomegranited** [comment]
Closed in favor of https://github.com/openedx/frontend-app-course-authoring/pull/655

**Pull request closed without merging**
